bouncer/src/repo/manipulator/modeloptimizer/repo_optimizer_multipart.cpp

```diff
@@ -84,207 +86,559 @@ bool MultipartOptimizer::collectMeshData(
-					{
-						for (uint32_t i = 0; i < submUVs.size(); ++i)
-						{
-							uvChannels[i].insert(uvChannels[i].end(), submUVs[i].begin(), submUVs[i].end());
-						}
-					}
-					else
-					{
-						//This shouldn't happen, if it does, then it means the mFormat isn't set correctly
-						repoError << "Unexpected transformedMesh format mismatch occured!";
-					}
-				}
-				else
-				{
-					repoError << "Failed merging meshes: Vertices or faces cannot be null!";
-				}
+		mapped.meshMapping.push_back(meshMap);
+
+		mapped.vertices.insert(mapped.vertices.end(), submVertices.begin(), submVertices.end());
+		for (const auto face : submFaces)
+		{
+			repo_face_t offsetFace;
+			for (const auto idx : face)
+			{
+				offsetFace.push_back(meshMap.vertFrom + idx);
 			}
-			break;
+			mapped.faces.push_back(offsetFace);
 		}
+
+		if (submNormals.size()) {
+			mapped.normals.insert(mapped.normals.end(), submNormals.begin(), submNormals.end());
+		}
+
+		if (submColors.size()) {
+			mapped.colors.insert(mapped.colors.end(), submColors.begin(), submColors.end());
+		}
+
+		if (mapped.uvChannels.size() == 0 && submUVs.size() != 0)
+		{
+			//initialise uvChannels
+			mapped.uvChannels.resize(submUVs.size());
+		}
+
+		if (mapped.uvChannels.size() == submUVs.size())
+		{
+			for (uint32_t i = 0; i < submUVs.size(); ++i)
+			{
+				mapped.uvChannels[i].insert(mapped.uvChannels[i].end(), submUVs[i].begin(), submUVs[i].end());
+			}
+		}
+		else
+		{
+			//This shouldn't happen, if it does, then it means the mFormat isn't set correctly
+			repoError << "Unexpected transformedMesh format mismatch occured!";
 		}
 	}
 	else
 	{
-		repoError << "Scene or node is null!";
+		repoError << "Failed merging meshes: Vertices or faces cannot be null!";
 	}
+}
 
-	return success;
+void updateMinMax(repo::lib::RepoVector3D& min, repo::lib::RepoVector3D& max, repo::lib::RepoVector3D v)
+{
+	if (v.x < min.x) {
+		min.x = v.x;
+	}
+	if (v.y < min.y) {
+		min.y = v.y;
+	}
+	if (v.z < min.z) {
+		min.z = v.z;
+	}
+
+	if (v.x > max.x) {
+		max.x = v.x;
+	}
+	if (v.y > max.y) {
+		max.y = v.y;
+	}
+	if (v.z > max.z) {
+		max.z = v.z;
+	}
 }
 
-repo::core::model::MeshNode* MultipartOptimizer::createSuperMesh
-(
-	const repo::core::model::RepoScene *scene,
-	const std::set<repo::lib::RepoUUID>           &meshGroup,
-	std::unordered_map<repo::lib::RepoUUID, repo::lib::RepoUUID, repo::lib::RepoUUIDHasher>  &matIDs,
-	const bool isGrouped)
+void MultipartOptimizer::splitMeshes(
+	const repo::core::model::RepoScene* scene,
+	repo::core::model::MeshNode node,
+	std::vector<mapped_mesh_t>& mappedMeshes
+)
 {
-	std::vector<repo::lib::RepoVector3D> vertices, normals;
-	std::vector<repo_face_t> faces;
-	std::vector<std::vector<repo::lib::RepoVector2D>> uvChannels;
-	std::vector<repo_color4d_t> colors;
-	std::vector<repo_mesh_mapping_t> meshMapping;
+	auto start = boost::chrono::high_resolution_clock::now();
 
-	repo::core::model::MeshNode* resultMesh = nullptr;
+	using Scalar = float;
```

Comment:
Since a couple of Bvh types take it as a generic parameter, I was thinking in case we wanted to move the whole thing over to doubles in the future.
(The Bvh and Vector3 types are defined immediately after, but the also the BoundingBox types use it below).
Perhaps as per your comment below it would be better to just define all the types at the top of the file without Scalar?

Reply:
> Perhaps as per your comment below it would be better to just define all the types at the top of the file without Scalar?
Could keep it that way if that's the intention just figured I'd ask as it's a bit odd. But yes I think we should define it at the top of the file